cvmfs/util/posix.cc

```diff
@@ -540,8 +545,15 @@ void ReadHalfPipe(int fd, void *buf, size_t nbyte) {
       SafeSleepMs(backoff_ms);
       if (backoff_ms < max_backoff_ms) backoff_ms *= 2;
     }
+  if (timeout_ms != 0) {
+    duration_ms = (platform_monotonic_time_ns() - timestamp) / 1000;
+    if (duration_ms  > timeout_ms)
+        return false;
+  }
   } while (num_bytes == 0);
-  assert((num_bytes >= 0) && (static_cast<size_t>(num_bytes) == nbyte));
+  if (timeout_ms == 0)
+    assert((num_bytes >= 0) && (static_cast<size_t>(num_bytes) == nbyte));
+  return ((num_bytes >= 0) && (static_cast<size_t>(num_bytes) == nbyte));
```

Comment:
At this point, read returned with some bytes or with an unexpected error. I think both short reads and unexpected errors have nothing to do with the timeout, so they should crash in any case.
```suggestion
  assert((num_bytes >= 0) && (static_cast<size_t>(num_bytes) == nbyte));
  return true;
```